File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> prefix(n), suffix(n);
        prefix[0] = height[0];
        suffix[n-1] = height[n-1];
        for(int i=1;i<n;i++){
            prefix[i] = max(prefix[i-1], height[i]);
        }
        for(int i=n-2;i>=0;i--){
            suffix[i] = max(suffix[i+1], height[i]);
        }
        int total = 0;
        for(int i=0;i<n;i++){
            if(prefix[i] > height[i] && suffix[i] >  height[i]){
                total += (min(prefix[i], suffix[i]) - height[i]);
            }
        }
        return total; 
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int l=0, h=n-1;
        int ans=0;
        int lmax=0, rmax=0;
        while(l<h){
            if(height[l] <= height[h]){
                if(lmax > height[l]){
                    ans += lmax - height[l];
                }
                else lmax = height[l];
                l++;
            }
            else{
                if(rmax > height[h]) ans += rmax - height[h];
                else rmax = height[h];
                h--;
            }
        }
        return ans;
    }
}